File: Semester_05/CIT-316(AI Sessional)/DFS.py
# def dfs(start, adj, n):
#     visited = [False] * (n + 1)
#     stack = [start]
    
#     while stack:
#         node = stack.pop()
        
#         if visited[node]:
#             continue
            
#         visited[node] = True
#         print(node, end=' ')
        
#         for i in range(len(adj[node]) - 1, -1, -1):
#             child = adj[node][i]
#             if not visited[child]:
#                 stack.append(child)

# n = 5
# adj = {
#     1: [2, 3],
#     2: [1, 4, 5],
#     3: [1],
#     4: [2],
#     5: [2]
# }

# print("DFS: ", end='')
# dfs(1, adj, n)



def dfs(node, adj, visited):
    visited.add(node)
    print(node, end=' ')
    
    for child in adj[node]:
        if child not in visited:
            dfs(child, adj, visited)

adj = {
    'B': ['A', 'C', 'D', 'E'],
    'A': ['C'],
    'C': ['G'],
    'D': ['E'],
    'E': ['F'],
    'F': ['I'],
    'G': ['H', 'I', 'J'],
    'H': ['K'],
    'I': ['J', 'L'],
    'J': ['K'],
    'K': [],
    'L': []
}

visited = set()
print("DFS (Recursive): ", end='')
dfs('B', adj, visited)
print()

File: Semester_05/CIT-316(AI Sessional)/DFS.cpp
// #include <bits/stdc++.h>
// using namespace std;

// void dfs(int node, int parent, vector<int> adj[]) 
// {
//     cout << node << " "; 

//     for (int child : adj[node]) 
//     {
//         if (child != parent) 
//         { 
//             dfs(child, node, adj); 
//         }
//     }
// }

// int main() 
// {
//     int n = 5;
//     vector<int> adj[n + 1];

//     adj[1] = {2, 3};
//     adj[2] = {1, 4, 5};
//     adj[3] = {1};
//     adj[4] = {2};
//     adj[5] = {2};

//     cout << "DFS (Recursive): ";
//     dfs(1, -1, adj); 

//     return 0;
// }


//DFS (Iterative Method) — using Stack

// #include <bits/stdc++.h>
// using namespace std;

// void dfs_iterative(int start, vector<int> adj[], int n) 
// {
//     vector<bool> visited(n + 1, false);
//     stack<int> st;
//     st.push(start);

//     while (!st.empty()) 
//     {
//         int node = st.top();
//         st.pop();

//         if (visited[node]) continue;
//         visited[node] = true;
//         cout << node << " ";

//         for (int i = adj[node].size() - 1; i >= 0; i--) 
//         {
//             int child = adj[node][i];
//             if (!visited[child])
//             {
//                 st.push(child);
//             }
//         }
//     }
// }

// int main() 
// {
//     int n = 5;
//     vector<int> adj[n + 1];

//     adj[1] = {2, 3};
//     adj[2] = {1, 4, 5};
//     adj[3] = {1};
//     adj[4] = {2};
//     adj[5] = {2};

//     cout << "DFS (Iterative): ";
//     dfs_iterative(1, adj, n);

//     return 0;
// }


#include <bits/stdc++.h>
using namespace std;

void dfs(char node, map<char, vector<char>> &adj, set<char> &visited) 
{
    visited.insert(node);
    cout << node << " ";

    for (char child : adj[node]) 
    {
        if (visited.find(child) == visited.end()) 
        {
            dfs(child, adj, visited);
        }
    }
}

int main() 
{
    map<char, vector<char>> adj;

    adj['B'] = {'A', 'C', 'D', 'E'};
    adj['A'] = {'C'};
    adj['C'] = {'G'};
    adj['D'] = {'E'};
    adj['E'] = {'F'};
    adj['F'] = {'I'};
    adj['G'] = {'H', 'I', 'J'};
    adj['H'] = {'K'};
    adj['I'] = {'J', 'L'};
    adj['J'] = {'K'};
    adj['L'] = {};

    set<char> visited;
    cout << "DFS (Recursive): ";
    dfs('B', adj, visited);
    cout << endl;

    return 0;
}